### ISSUE TITLE
'class N2kStream' has no member named 'peek'

### ISSUE THREAD

**Issue by peterantypas** [ttlappalainen/NMEA2000#225]
This class does not define peek() in its non-Arduino implementation. Is it supposed to return int like read(), but without advancing the read pointer?

    #ifdef ARDUINO
    // Arduino users get away with using the standard Stream class and its
    // subclasses. Forward declare the Stream class here and include Arduino.h in
    // the application.
    #include <Arduino.h>
    typedef Stream N2kStream;
    #else
    // Non Arduino platforms need to implement this themselves if they want to use
    // functions which operate on streams.
    class N2kStream {
       public:
       // Returns first byte if incoming data, or -1 on no available data.
       virtual int read() = 0;
    
       // Write data to stream.
       virtual size_t write(const uint8_t* data, size_t size) = 0;
    
       // Print string to stream.
       size_t print(const char* str);
    
    #if defined(__AVR__)
       // Flash stored string stream for AVR platforms.
       size_t print(const __FlashStringHelper* str);
       size_t println(const __FlashStringHelper* str);
    #endif
    
       // Print value to stream.
       size_t print(int val, uint8_t radix = 10);
    
       // Print string and newline to stream.
       size_t println(const char *str);
    
       // Print value and newline to stream.
       size_t println(int val, uint8_t radix = 10);
    };
    #endif

**Comment by peterantypas** [comment]
This is being called from tActisenseReader::GetMessageFromStream(), possibly others as well.

**Comment by ttlappalainen** [comment]
Yes. It just reads one char from stream, but does not take it away.

**Comment by peterantypas** [comment]
Where is the current implementation of read()?  It's  a pure virtual here.

**Comment by ttlappalainen** [comment]
For RPi there is tSocketStream defined in NMEA2000_socketCAN.

**Comment by peterantypas** [comment]
So I have to write my own implementation for STM32, right?

**Comment by peterantypas** [comment]
I think I answered my own question, so yes, I'll have to write one. That said, the existing code will not compile outside Arduino because N2kStream does not define peek() even as a pure virtual.

**Comment by ttlappalainen** [comment]
I think no. I modified tActisenseReader::GetMessageFromStream() some time ago with that function to allow multi protocol handling for stream. If you do not need to handle Actisense format data, you do not need that module. I'll have to check and add peek as virtual function for N2kStream and/or maybe add define so that it does not use peek.
